resolwe/flow/models.py

```diff
@@ -479,6 +479,35 @@ def add_dependency(value):
                 for data in value:
                     add_dependency(data)
 
+    def validate_json_fields(self):
+        try:
+            if self.pk is None:  # on create
+                validate_schema(self.input, self.process.input_schema)  # pylint: disable=no-member
+
+            if self.descriptor_schema:
+                validate_schema(self.descriptor, self.descriptor_schema.schema)  # pylint: disable=no-member
+            elif self.descriptor and self.descriptor != {}:
+                raise ValueError("`descriptor_schema` must be defined if `descriptor` is given")
+
+            path_prefix = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(self.pk))
+            output_schema = self.process.output_schema  # pylint: disable=no-member
+            if self.status == Data.STATUS_DONE:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix)
+            else:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix,
+                                test_required=False)
+        except ValidationError as ex:
+            # if object doesn't have ``pk`` assigned and doesn't validate, it
+            # won't be stored in database, so we don't have to do anything
+            if self.pk:
+                # current object is already corupted, so we have to get fresh
+                # copy (last that has passed validation)
+                data = Data.objects.get(pk=self.pk)
```

Comment:
I agree. Please move the error handling to the process executor. The model should raise ValidationError. I will give an example where the current solution behaves strange. Say you wish to change the data object directly on the server in the Django shell. You make a typo in annotations and run `save()`. Instead of getting the ValidationError, the data object would be saved with an ERROR status.